### PR TITLE
Adjust pessimistic lock protocol for lock behavior optimizations

### PR DESCRIPTION
**Pull request by MyonKeminta**
This is part of an attempt to demo an optimization about pessimistic lock waking up behavior.

### CODE REVIEW

**Comment by MyonKeminta** [comment]
@sticnarf @cfzjywxk PTAL
It would be good if you have any suggestions about naming 😥

File: proto/kvrpcpb.proto [hunk 1]
```diff
@@ -173,6 +178,29 @@ message PessimisticLockRequest {
...
+    bytes value = 2;
+    bool existence = 3;
+    // We allow a key be locked when there is write conflict (latest commit_ts > for_update_ts).
+    // In this case, the key is semantically locked by a newer for_update_ts.
+    // For each requested key, non-zero if the key is locked with write conflict, and the
+    // value is the commit_ts of the latest version of the specified key, which is the same value
+    // as the actual for_update_ts. At the sametime, the corresponsing value in `values` field and
+    // `not_founds` field will be the latest value of the key even if `return_values` and
+    // `check_existence` are not set.
+    uint64 locked_with_conflict_ts = 4;
```
Comment (cfzjywxk) [inline proto/kvrpcpb.proto:203]: > is the commit_ts of the latest version of the specified key, which is the same value as the actual for_update_ts.

I'm confused about this sentence. If the key is locked with conflict and the `commit_ts` of the key is larger than the try-lock transaction `for_update_ts`, the `locked_with_conflict_ts` would be set to the latest `commit_ts`?

File: proto/kvrpcpb.proto [hunk 2]
```diff
@@ -137,6 +137,11 @@ message PrewriteResponse {
     ExecDetailsV2 exec_details_v2 = 5;
 }
 
+enum PessimisticLockWaitingMode {
+    RetryAfterWait = 0;
+    ResumeAfterWait = 1;
+}
```
[the remark this answers is not in the document]
Comment (cfzjywxk) [reply proto/kvrpcpb.proto:143]: How about using `Normal` for the original mode,  and `ForceLock` or `LockBeforeRetry` for the new mode?

[the remark this answers is not in the document]
Comment (MyonKeminta) [reply proto/kvrpcpb.proto:143]: How about
```proto
enum PessimisticLockWakeUpMode {
    Normal = 0;
    ForceLock = 1;
}
```

File: proto/kvrpcpb.proto (hunk 1 above)
Comment (MyonKeminta) [reply proto/kvrpcpb.proto:203]: Yes. I adjusted the expression of the comment.

File: proto/kvrpcpb.proto (hunk 2 above)
[the remark this answers is not in the document]
Comment (cfzjywxk) [reply proto/kvrpcpb.proto:143]: LGTM, we could add some comments here too.

File: proto/kvrpcpb.proto [hunk 3]
```diff
@@ -859,6 +901,9 @@ message LockInfo {
     bool use_async_commit = 8;
     uint64 min_commit_ts = 9;
     repeated bytes secondaries = 10;
+    // Hint the client that resolving lock is not needed for this lock. For `PessimisticLock`
+    // requests only.
+    bool skip_resolving_lock = 11;
```
Comment (sticnarf) [inline proto/kvrpcpb.proto:906]: Could it be put in `PessimisticLockKeyResult`? I think it's better not to pollute `LockInfo` because of one request.

**Review by cfzjywxk — APPROVED** [review]

**Review by sticnarf — APPROVED** [review]

**Pull request merged**